### PR TITLE
[fixes #1849] - Keycloak Extension

### LINKED ISSUE #1849: Keycloak Extension
This issue is about providing an extension to protect Quarkus applications using Keycloak. Keycloak provides "adapters" for different platforms and servlet containers, including Undertow. 

This extension should leverage the existing integration with Undertow in Keycloak and provide the same capabilities just like any other Keycloak adapter.

### PR DESCRIPTION
**Pull request by pedroigor**
Please, do not merge. These changes rely on a Keycloak release. Sending it now to gather feedback about the changes.

### CODE REVIEW

**Comment by kenfinnigan** [comment]
Haven't looked in detail at the code but I have a few questions.

How does this relate, or integrate, with MP JWT? Does this create two ways of doing declarative security in applications? Is that a good thing?

Also, we prefer to have all configuration in `application.properties` so I'd suggest having config properties for the content that is required in `keycloak.json` and setting them into Keycloak Client as needed.

File: docs/src/main/asciidoc/keycloak-guide.adoc [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+[source,bash]
+----
+docker run --name keycloak -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin -p 8180:8080 quay.io/keycloak/keycloak
+----
+
+You should be able to access your Keycloak Server at http://localhost:8180/auth[localhost:8180/auth].
+
+Log in as the `admin` user to access the Keycloak Administration Console. Username should be `admin` and password `admin`.
+
+Import the {quickstarts-tree-url}/using-keycloak/config/quarkus-realm.json[realm configuration file] to create a new realm.
```
Comment (sebastienblanc) [inline docs/src/main/asciidoc/keycloak-guide.adoc]: Maybe a link to our doc or a screenshot to explain how to import a realm ?

File: devtools/common/src/main/filtered/extensions.json [hunk 2]
```diff
@@ -403,5 +403,15 @@
     ],
     "groupId": "io.quarkus",
     "artifactId": "quarkus-flyway"
+  },
+  {
+    "name": "Keycloak",
```
Comment (machi1990) [inline devtools/common/src/main/filtered/extensions.json]: I think we can add `Keycloak` extensions by following alphabetical order as suggested by this [PR](https://github.com/quarkusio/quarkus/pull/1856)

**Comment by kenfinnigan** [comment]
And I'm not saying we don't want to use Keycloak, it's more a case of figuring out what the security "story" is for Quarkus

**Comment by pedroigor** [comment]
> Haven't looked in detail at the code but I have a few questions.
> 
> How does this relate, or integrate, with MP JWT? Does this create two ways of doing declarative security in applications? Is that a good thing?

Good point @kenfinnigan. However, there is a difference that makes the two approaches complementary.

The overlap exists until need something beyond JWT and, as a consequence, MP-JWT. Although Keycloak is issuing tokens using the JWT format, we provide much more than what JWT can offer. Keycloak is an OpenID Connect Provider and OAuth 2.0 compliant Authorization Server and, as such, there are other standards, implementations and use cases involved. That said, the extension herein is targeted for those looking for bearer token authorization based on OpenID Connect and OAuth 2.0.

In addition to that, Keycloak also provides some caveats and implementations that leverage both specifications. We are constantly involving to provide to our users a good experience when protecting their apps with those standards. Some of these improvements and how to use them don't fit into a standard JWT implementation. 

As a matter of fact, developers should consider this extension if they want to enable OAuth 2.0 to their applications, backed by Keycloak. However, the other way around is valid where you may want to use Keycloak as an IAM solution and MP-JWT to integrate with Keycloak. Keycloak tokens are basically JWTs and their data structure (claims) can be changed accordingly with your requirements. For instance, a new feature was added to Keycloak to improve the user experience for those using MP-JWT so that Keycloak can issue MP-JWT compliant tokens.

> Also, we prefer to have all configuration in `application.properties` so I'd suggest having config properties for the content that is required in `keycloak.json` and setting them into Keycloak Client as needed.

I will work on that. One question though. We already have in Keycloak a specific type that maps the different configuration options we need. It seems that in Quarkus I would need to redefine these properties to annotate them with `@ConfigItem`. Is that correct or there is some way to automatically map properties using reflection?

**Comment by sberyozkin** [comment]
Hi @pedroigor
You are right Keycloak can have many usecases/flows supported, but from 'the point of view' of the Quarkus endpoint, once the token is injected into (or somehow accessed from) its JAX-RS endpoint, it is just JWT, so it would be nice to be able to access this token with MP JWT API.
May be a KC principal can be additionally made available as MP JWT Principal ? We can easily enough introduce some smalltye-jwt interface or some hook to tap into the KC adapter validation flow...
What do you think ?  Can we consider it during the phase2 :-) ?

The other question, If the Quarkus JAX-RS endpoint is in fact an 'OIDC client' endpoint, will this extension allow accessing the IdTokens (not only access tokens) from the endpoints ? 

Thanks

**Comment by pedroigor** [comment]
Hello @sberyozkin,

> You are right Keycloak can have many usecases/flows supported, but from 'the point of view' of the Quarkus endpoint, once the token is injected into (or somehow accessed from) its JAX-RS endpoint, it is just JWT, so it would be nice to be able to access this token with MP JWT API.
> May be a KC principal can be additionally made available as MP JWT Principal ? We can easily enough introduce some smalltye-jwt interface or some hook to tap into the KC adapter validation flow...
> What do you think ? Can we consider it during the phase2 :-) ?

Sure. We can definitely have that. If you look at the guide for this extension, you'll see that developers can inject a `KeycloakSecurityContext` object, which is quite used by applications using Keycloak. If we could also allow people to inject a  JsonWebToken/Principal backed by this `KeycloakSecurityContext` object, it would be great.

> 
> The other question, If the Quarkus JAX-RS endpoint is in fact an 'OIDC client' endpoint, will this extension allow accessing the IdTokens (not only access tokens) from the endpoints ?

In theory, yes. At the moment I'm only considering authorization given that microprofile is mainly targeted for implementing backends using JAX-RS. I'm not providing tests for scenarios where the application is doing the Hybrid Flow.

> 
> Thanks

**Comment by sberyozkin** [comment]
@pedroigor, thanks. I'll have a look at some point after this extension is in on how a link it with MP-JWT might be established, and ask for your advise along the way :-)
Re IdToken - I was thinking more about the authorization code OIDC clients which run as the servers (never understood the hybrid flow well :-) ), but indeed, it is not of the immediate interest

Thanks

**Comment by pedroigor** [comment]
I'm still writing the config items so we can use `application.properties` for the adapter configuration.

**Comment by starksm64** [comment]
> I will work on that. One question though. We already have in Keycloak a specific type that maps the different configuration options we need. It seems that in Quarkus I would need to redefine these properties to annotate them with `@ConfigItem`. Is that correct or there is some way to automatically map properties using reflection?

The current capabilities do require a configuration object annotated with @ConfigRoot and @Configitem. You would then have to consume this in a processor build step that output a build item containing your KC specific configuration object that the the runtime template would use to configuration the runtime deployment.

**Comment by starksm64** [comment]
There are overlaps and there are differences. The smallrye-jwt extension is supporting injection of the MP-JWT spec beans. The keycloak extension is supporting the external KC identity manager which is a requirement for cloud deployments. Perhaps these could be merged in the future if configuration selection allows for removal of unused behaviors in the final images.

**Comment by pedroigor** [comment]
@starksm64, I manage to write the configuration objects. However, some of the settings we have in `keycloak.json` should be mapped as a list of objects which I could not figure out how to do this mapping using MP-Config. For instance:

```json
quarkus.keycloak.policy-enforcer.paths[0].path=/api/a
quarkus.keycloak.policy-enforcer.paths[0].enforcement-mode=DISABLED
quarkus.keycloak.policy-enforcer.paths[1].path=/api/b
quarkus.keycloak.policy-enforcer.paths[1].enforcement-mode=PERMISSIVE
```

Where `paths` maps to a config item that is a list of path objects. Any idea how to achieve this?

**Comment by starksm64** [comment]
> Where `paths` maps to a config item that is a list of path objects. Any idea how to achieve this?

We don't support a List of objects out of the box, so you would either have to write an MP config converter, or use a map of objects and use the index as the key. For example, you could create a `PolicyEnforcement` config group that has the path and enforcement mode, and then include a `Map<String, PolicyEnforcement>` in your config root:

```java
@ConfigGroup
public class PolicyEnforcement {
    /** PE path */
    @ConfigItem
    String path;
    /** The KC enforcement-mode */
    @ConfigItem(defaultValue = "PERMISSIVE")
    public String enforcementMode;
}

@ConfigRoot(name = "keycloak")
public class KCConfig {
    /** A collection of policy-enforcements */
    @ConfigItem
    Map<String, PolicyEnforcement> policyEnforcer;
...
}
```
The corresponding configuration would be:

```properties
quarkus.keycloak.policy-enforcer.0.path=/api/a
quarkus.keycloak.policy-enforcer.0.enforcement-mode=DISABLED
quarkus.keycloak.policy-enforcer.1.path=/api/b
quarkus.keycloak.policy-enforcer.1.enforcement-mode=PERMISSIVE
```

**Comment by pedroigor** [comment]
@starksm64, PR updated with changes to configuration. Tests also added. Thanks for the tips.

**Comment by pedroigor** [comment]
@starksm64, PR updated with latest Keycloak release (6.0.0).

**Comment by pedroigor** [comment]
Quickstart https://github.com/quarkusio/quarkus-quickstarts/pull/161.

**Comment by starksm64** [comment]
I'm looking into the current failures

**Comment by pedroigor** [comment]
@starksm64, thanks for the fix.

**Review by starksm64 — APPROVED** [review]
I have been through the quickstart and verified it against the extension.

**Pull request merged**